common/datablocks/src/kernels/data_block_group_by.rs

```diff
@@ -100,14 +89,214 @@ impl DataBlock {
+            HashMethodKind::KeysU64(_) => DataType::UInt64,
+        }
+    }
+}
+
+impl DataBlock {
+    pub fn choose_hash_method(
+        block: &DataBlock,
+        column_names: &[String],
+    ) -> Result<HashMethodKind> {
+        let mut group_key_len = 0;
+        for col in column_names {
+            let column = block.try_column_by_name(col)?;
+            let typ = column.data_type();
+            if common_datavalues::is_integer(&typ) {
+                group_key_len += common_datavalues::numeric_byte_size(&typ)?;
+            } else {
+                return Ok(HashMethodKind::Serializer(HashMethodSerializer::default()));
+            }
+        }
+        match group_key_len {
+            1 => Ok(HashMethodKind::KeysU8(HashMethodKeysU8::default())),
+            2 => Ok(HashMethodKind::KeysU16(HashMethodKeysU16::default())),
+            3..=4 => Ok(HashMethodKind::KeysU32(HashMethodKeysU32::default())),
+            5..=8 => Ok(HashMethodKind::KeysU64(HashMethodKeysU64::default())),
+            _ => Ok(HashMethodKind::Serializer(HashMethodSerializer::default())),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct HashMethodSerializer {}
+
+impl HashMethodSerializer {
+    #[inline]
+    pub fn get_key(&self, array: &DFBinaryArray, row: usize) -> Vec<u8> {
+        let v = array.as_ref().value(row);
+        v.to_owned()
+    }
+}
+impl HashMethod for HashMethodSerializer {
+    type HashKey = Vec<u8>;
+
+    fn build_keys(&self, group_columns: &[&DataColumn], rows: usize) -> Result<Vec<Self::HashKey>> {
+        let mut group_keys = Vec::with_capacity(rows);
+        {
+            let mut group_key_len = 0;
+            for col in group_columns {
+                let typ = col.data_type();
+                if common_datavalues::is_integer(&typ) {
+                    group_key_len += common_datavalues::numeric_byte_size(&typ)?;
+                } else {
+                    group_key_len += 4;
+                }
+            }
+
+            for _i in 0..rows {
+                group_keys.push(Vec::with_capacity(group_key_len));
+            }
+
+            for col in group_columns {
+                DataColumn::serialize(col, &mut group_keys)?;
+            }
+        }
+        Ok(group_keys)
+    }
+}
+
+macro_rules! build_primitive_keys {
+    ($group_columns: ident, $rows: ident) => {{
+        let mut group_keys = vec![0; $rows];
+        let mut offsize = 0;
+        let ptr = group_keys.as_mut_ptr() as *mut u8;
+        let step = std::mem::size_of::<Self::HashKey>();
+
+        let mut size = step;
+        while size > 0 {
+            build(size, &mut offsize, $group_columns, ptr, step)?;
+            size /= 2;
+        }
+
+        Ok(group_keys)
+    }};
+}
+#[derive(Debug, Clone, Default)]
+
+pub struct HashMethodKeysU8 {}
+impl HashMethodKeysU8 {
+    #[inline]
+    pub fn get_key(&self, array: &DFUInt8Array, row: usize) -> u8 {
+        array.as_ref().value(row)
```

Comment:
what happen when value is null?

Reply:
Can we distinguish between zero and null?

Reply:
No, currently null is not taken into consideration in the group by.